### PR TITLE
feat: DHIS2-10161 [part 02] Create enrollment page with routing | fetches enrollments

### PR DESCRIPTION
**Pull request by paschalidi**
### What is this PR about?

Basically we are fetching the enrollments and storing them on redux. This PR works out the case that the user types in the url the all the parameters needed. In other words this will work _only_ for urls that provide all the following information http://localhost:3000/#/enrollment?programId=uy2gU8kT1jF&orgUnitId=DiszpKrYNg8&teiId=EaOyKGOIGRp&enrollmentId=k8ZtDibWX5Y 

In the next PRs we will introduce the usecases where you can load the whole page using only the `enrollmentId` but for now this is out of the scope of _this_ PR.

Expect to see something between these lines when you run this code locally. 

### The next PR. 

The next pr will add styling in the locked selector and also include the persons name _and_ the enrollment id in the locked selector.

### CODE REVIEW

File: src/reducers/descriptions/trackerCapture.reducerDescriptions.js [hunk 1]
```diff
@@ -74,16 +74,15 @@ import { currentSelectionsReducerDesc } from './currentSelections.reducerDescrip
 import { feedbackDesc } from './feedback.reducerDescription';
 import { activePageDesc } from '../../core_modules/capture-core/reducers/descriptions/activePage.reducerDescription';
 import { searchPageDesc } from '../../core_modules/capture-core/reducers/descriptions/searchPage.reducerDescription';
+import { enrollmentPageDesc } from '../../core_modules/capture-core/reducers/descriptions/enrollmentPage.reducerDescription';
 import { newPageDesc } from '../../core_modules/capture-core/reducers/descriptions/newPage.reducerDescription';
 
 export default [
```
Comment (paschalidi) [inline src/reducers/descriptions/trackerCapture.reducerDescriptions.js:80]: I ordered alphabetically here, since it was already there I thought is good if we have it complete. 
Sorry this is out of scope for this PR.

File: src/core_modules/capture-core/components/Pages/Enrollment/EnrollmentPage.epics.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+                                        moment.utc(a.enrollmentDate).diff(moment.utc(b.enrollmentDate)));
+
+                                    return successfulFetchingEnrollmentPageInformationFromUrl({
+                                        selectedName,
+                                        enrollmentsSortedByDate,
+                                    });
+                                }),
+                            )),
+                    startWith(showLoadingViewOnEnrollmentPage()),
+                    catchError(() => of(showErrorViewOnEnrollmentPage())),
```
Comment (paschalidi) [inline src/core_modules/capture-core/components/Pages/Enrollment/EnrollmentPage.epics.js:41]: This bit will be reworked a lot in the future PRs. Perhaps it doesnt look so amazing in this one, but I promise I will be making it look better when we also start fetching these information using the `teiId`. Now we only fetch using the `enrollmentId`

File: src/core_modules/capture-core/components/Pages/Enrollment/EnrollmentPage.epics.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+import { getApi } from '../../../d2';
+import {
+    enrollmentPageActionTypes,
+    showErrorViewOnEnrollmentPage,
+    showLoadingViewOnEnrollmentPage,
+    successfulFetchingEnrollmentPageInformationFromUrl,
+} from './EnrollmentPage.actions';
+
+const fetchEnrollment = id => getApi().get(`enrollments/${id}`, { fields: 'trackedEntityInstance,program,orgUnit' });
+const fetchTrackedEntityInstance = id => getApi().get(`trackedEntityInstances/${id}`, { fields: 'attributes,enrollments' });
```
Comment (JoakimSM) [inline src/core_modules/capture-core/components/Pages/Enrollment/EnrollmentPage.epics.js:15]: Reminder to use the data engine here instead

File: src/core_modules/capture-core/components/Pages/Enrollment/EnrollmentPage.container.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+export const EnrollmentPage: ComponentType<{||}> = () => {
+    const dispatch = useDispatch();
+
+    const enrollmentId: string =
+      useSelector(({ router: { location: { query } } }) => query.enrollmentId);
+    const enrollments: Object =
+      useSelector(({ enrollmentPage }) => enrollmentPage.enrollments);
+
+    useEffect(() => {
+        const enrollmentIsNotPreset = !(enrollments && enrollments.some(({ enrollment }) => enrollmentId === enrollment));
```
Comment (JoakimSM) [inline src/core_modules/capture-core/components/Pages/Enrollment/EnrollmentPage.container.js:18]: probably dealt with later, but I see a loop forming if you query an enrollmentId that doesn't exist.

**Review by JoakimSM — COMMENTED** [review]
Looks good 👍 Would be perfect if you use the data engine instead of the old d2 library.

File: src/core_modules/capture-core/components/Pages/Enrollment/EnrollmentPage.epics.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        flatMap(() => {
+            const { query: { enrollmentId } } = store.value.router.location;
+
+            return from(fetchEnrollment(enrollmentId))
+                .pipe(
+                    flatMap(({ trackedEntityInstance }) =>
+                        from(fetchTrackedEntityInstance(trackedEntityInstance))
+                            .pipe(
+                                map(({ attributes, enrollments }) => {
+                                    const selectedName = attributes.reduce((acc, { value: dataElementValue }) =>
```
Comment (JoakimSM) [inline src/core_modules/capture-core/components/Pages/Enrollment/EnrollmentPage.epics.js:29]: We need to use the metadata when computing the name. We should display the attribute value only if the "display in list" flag is set. We should also use the attribute order from the metadata.

Comment (paschalidi) [reply src/core_modules/capture-core/components/Pages/Enrollment/EnrollmentPage.epics.js:29]: Okay I was wondering why I can see that display on list flag. Then I thought that it might have been that on the backend they return only the values that are good for display.. 

Okay then you mean that we should be using the data we have stored in memory right?

Comment (JoakimSM) [reply src/core_modules/capture-core/components/Pages/Enrollment/EnrollmentPage.epics.js:29]: Yes 👍

**Review by JoakimSM — APPROVED** [review]
